File: libflexlay/gui/editor_map_widget.cpp
// Flexlay - A Generic 2D Game Editor
// Copyright (C) 2014 Ingo Ruhnke <grumbel@gmail.com>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

#include "gui/editor_map_widget.hpp"

#include <QDropEvent>
#include <QMouseEvent>
#include <QPainter>
#include <QWidget>
#include <QMimeData>

#include "editor_map.hpp"

EditorMapWidget::EditorMapWidget(EditorMapComponent& comp, QWidget* parent) :
  QWidget(parent),
  m_comp(comp)
{
  QPalette pal = palette();
  pal.setColor(backgroundRole(), QColor(100, 80, 100));
  setPalette(pal);

  setAutoFillBackground(true);
  setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
  setAcceptDrops(true);
}

EditorMapWidget::~EditorMapWidget()
{
}

void
EditorMapWidget::dragEnterEvent(QDragEnterEvent* event)
{
  std::cout << "dragEnter: " << event->mimeData()->hasFormat("application/supertux-badguy") << std::endl;
  if (event->mimeData()->hasFormat("application/supertux-badguy"))
  {
    event->accept();
  }
}

void
EditorMapWidget::dragLeaveEvent(QDragLeaveEvent* event)
{
  std::cout << "dragLeave" << std::endl;
}

void
EditorMapWidget::dropEvent(QDropEvent* event)
{
  std::cout << "drop happened: " << event->pos().x() << " " << event->pos().y() << std::endl;
  QByteArray data = event->mimeData()->data("application/supertux-badguy");
  int i = 0;
  memcpy(&i, data.data(), sizeof(i));
}

QSize
EditorMapWidget::sizeHint() const
{
  return QSize(1280, 800);
}

void
EditorMapWidget::on_map_change()
{
  std::cout << "EditorMapWidget::on_map_change()" << std::endl;
  Workspace workspace = m_comp.get_workspace();
  if (!workspace.is_null())
  {
    if (workspace.get_map() && workspace.get_map().has_bounding_rect())
    {
      Rect rect = workspace.get_map().get_bounding_rect();
      std::cout << "Setting minimum Size: " << rect.get_width() << " " << rect.get_height() << std::endl;
      setMinimumSize(QSize(rect.get_width(), rect.get_height()));

      QPalette pal = palette();
      pal.setColor(backgroundRole(), workspace.get_map().get_background_color().to_qt());
      setPalette(pal);

      repaint();
    }
  }
}

void
EditorMapWidget::mouseMoveEvent(QMouseEvent* event)
{
  Workspace workspace = m_comp.get_workspace();
  InputEvent ev(*event);
  workspace.mouse_move(ev);
  repaint();
}

void
EditorMapWidget::mousePressEvent(QMouseEvent* event)
{
  Workspace workspace = m_comp.get_workspace();
  InputEvent ev(*event);
  workspace.mouse_down(ev);
  std::cout << "mouse press: " << std::endl;
  repaint();
}

void
EditorMapWidget::mouseReleaseEvent(QMouseEvent* event)
{
  Workspace workspace = m_comp.get_workspace();
  InputEvent ev(*event);
  workspace.mouse_up(ev);
  std::cout << "mouse release: " << std::endl;
  repaint();
}

void
EditorMapWidget::paintEvent(QPaintEvent* event)
{
  QPainter painter;
  painter.begin(this);

  //painter.setRenderHint(QPainter::Antialiasing);

  Workspace workspace = m_comp.get_workspace();
  GraphicContext gc(m_comp.get_gc_state(), painter);
  m_comp.get_gc_state().push(gc);
  workspace.draw(gc);
  m_comp.get_gc_state().pop(gc);

  painter.end();
}

void
EditorMapWidget::resizeEvent(QResizeEvent* event)
{
  m_comp.get_gc_state().set_size(event->size().width(), event->size().height());
}

/* EOF */

File: flexlay/gui/editor_map_widget.py
# Flexlay - A Generic 2D Game Editor
# Copyright (C) 2014 Ingo Ruhnke <grumbel@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.


from typing import Optional, TYPE_CHECKING

from PyQt5.QtCore import QSize
from PyQt5.QtGui import (QColor, QPainter, QDragEnterEvent,
                         QDragLeaveEvent, QDropEvent,
                         QResizeEvent, QPaintEvent, QMouseEvent,
                         QWheelEvent)
from PyQt5.QtWidgets import QSizePolicy, QWidget  # QOpenGLWidget

from flexlay.input_event import InputEvent
from flexlay.graphic_context import GraphicContext
from flexlay.math import Pointf
from flexlay.util.signal import Signal

if TYPE_CHECKING:
    from flexlay.gui.editor_map_component import EditorMapComponent


class EditorMapWidget(QWidget):  # QOpenGLWidget

    def __init__(self, comp: 'EditorMapComponent', parent: Optional[QWidget]) -> None:
        super().__init__(parent)
        self.setMouseTracking(True)

        self.comp = comp
        self.workspace = comp.get_workspace()
        self.gc_state = comp.get_gc_state()
        self.painter = QPainter()
        self.gc = GraphicContext(self.painter, self.gc_state)

        pal = self.palette()
        pal.setColor(self.backgroundRole(), QColor(100, 0, 100))
        self.setPalette(pal)

        self.setAutoFillBackground(True)
        self.setSizePolicy(QSizePolicy.Expanding, QSizePolicy.Expanding)
        self.setAcceptDrops(True)
        self.sig_drop = Signal()
        self.sig_mouse_move = Signal()

    def dragEnterEvent(self, event: QDragEnterEvent) -> None:
        if event.mimeData().hasFormat("application/x-supertux-badguy"):
            event.accept()

    def dragLeaveEvent(self, event: QDragLeaveEvent) -> None:
        pass

    def dropEvent(self, event: QDropEvent) -> None:
        data = event.mimeData().data("application/x-supertux-badguy")
        pos = self.comp.screen2world(Pointf.from_qt(event.pos()))
        self.sig_drop(data, pos)

    def sizeHint(self) -> QSize:
        return QSize(1280, 800)

    def on_map_change(self) -> None:
        workspace = self.comp.get_workspace()
        if workspace and workspace.get_map():
            if workspace.get_map().has_bounding_rect():
                rect = workspace.get_map().get_bounding_rect()
                self.setMinimumSize(QSize(int(rect.width), int(rect.height)))

                pal = self.palette()
                pal.setColor(self.backgroundRole(), workspace.get_map().get_background_color().to_qt())
                self.setPalette(pal)

                self.repaint()

    def wheelEvent(self, event: QWheelEvent) -> None:
        num_degrees = event.angleDelta() / 8
        num_steps = int(num_degrees.manhattanLength() / 15)

        for _ in range(num_steps):
            if num_degrees.x() > 0 or num_degrees.y() > 0:
                self.comp.zoom_in(Pointf.from_qt(event.pos()))
            else:
                self.comp.zoom_out(Pointf.from_qt(event.pos()))

    def mouseMoveEvent(self, event: QMouseEvent) -> None:
        workspace = self.comp.get_workspace()
        ev = InputEvent.from_qt(event)
        workspace.mouse_move(ev)
        self.repaint()

        pos = self.comp.screen2world(Pointf.from_qt(event.pos()))
        self.sig_mouse_move(pos)

    def mousePressEvent(self, event: QMouseEvent) -> None:
        workspace = self.comp.get_workspace()
        ev = InputEvent.from_qt(event)
        workspace.mouse_down(ev)
        self.repaint()

    def mouseReleaseEvent(self, event: QMouseEvent) -> None:
        workspace = self.comp.get_workspace()
        ev = InputEvent.from_qt(event)
        workspace.mouse_up(ev)
        self.repaint()

    def paintEvent(self, event: QPaintEvent) -> None:
        self.painter.begin(self)

        # self.painter.setRenderHint(QPainter::Antialiasing)
        self.gc_state.push(self.gc)
        self.workspace.draw(self.gc)
        self.gc_state.pop(self.gc)

        self.painter.end()

    def resizeEvent(self, event: QResizeEvent) -> None:
        self.comp.get_gc_state().set_size(event.size().width(), event.size().height())


# EOF #
